fix(utxo): deserialize sapling root for PIVX block headers

shamardy opened the pull request:
This PR resolves a deserialization failure for `PIVX` block headers that was causing an `UnexpectedEnd` error.

The root cause was that the `BlockHeader` deserializer did not account for the `hash_final_sapling_root` field, which is present in all current `PIVX` headers. This caused the stream reader to consume an incorrect number of bytes, leading to a misaligned stream for subsequent headers.

The fix introduces a `PIVX` variant to `CoinVariant` and updates the `BlockHeader::deserialize` function to correctly read the `hash_final_sapling_root` field when the coin is identified as `PIVX`, ensuring proper headers parsing.

fixes https://github.com/KomodoPlatform/komodo-defi-framework/issues/2048#issuecomment-3152520745

shamardy committed:
fix(utxo): Correct PIVX block header deserialization

Updates the BlockHeader deserializer to correctly read the `hash_final_sapling_root` field for PIVX.

cipig reviewed (approved):
thanks for the fix
calling recover_funds_of_swap with the same uuid that previously failed with "UnexpectedEnd" now shows
` "error" : "rpc:198] RPC call failed: lp_swap:1519] saved_swap:130] maker_swap:1700] Too early to refund, wait until 1754357000"` which is in 1h

cipig commented:
oh, just saw that the swap is from 3 months ago and refund already failed with
```
      {
         "event" : {
            "data" : {
               "error" : "maker_swap:1315] !maker_coin.send_maker_refunds_payment: utxo_common:1958] HTLC spend fee 100000 is greater than transaction output 70000"
            },
            "type" : "MakerPaymentRefundFailed"
         },
         "timestamp" : 1747550467167
      },
      {
         "event" : {
            "type" : "Finished"
         },
         "timestamp" : 1747550467172
      }
```
so why does recover say i need to wait?
every time i call recover_funds_of_swap, the timestamp in the error increases

shamardy commented:
> every time i call recover_funds_of_swap, the timestamp in the error increases

Thanks for finding this. The deserializer assumes a Zcash-style field order, but PIVX places its `hash_final_sapling_root` field at a different position. This means the bytes fit and deserialization completes without errors, but the data assigned to the fields is corrupted. The bytes for the timestamp are incorrectly read as the sapling root, the bytes for the difficulty are read as the timestamp, and so on. Will fix it.

shamardy committed:
fix(deserialization): correct PIVX block header serde for Sapling fields

PIVX block headers with `nVersion` >= 8 include a `hash_final_sapling_root`, but its position in the byte stream is different from other Zcash-style chains. It appears after the `nNonce`, not before `nTime`.

The previous deserialization logic incorrectly assumed the Zcash-style order, leading to data corruption for all subsequent fields (`nTime`, `nBits`, etc.) and causing failures in functions that rely on block timestamps, such as swap refunds.

This commit corrects both the serialization and deserialization implementations for `BlockHeader` to handle the PIVX-specific field order. It also introduces a `pivx_mtp` test to validate the fix and prevent regressions.

shamardy commented:
should be fixed @cipig

cipig commented:
it now refunded that tx that was supposed to be unrefundable because of `HTLC spend fee 100000 is greater than transaction output 70000` 3 months ago... that's likely due to other changes to kdf in the last 3 months
```
be42d8c0-127c-4df5-800b-4de8318e1bf5
{
   "result" : {
      "action" : "RefundedMyPayment",
      "coin" : "PIVX",
      "tx_hash" : "2d69c282ed1f59ade4f0d4f26ef2cdfa67aef29ad728835fddcf1424c27cd316",
      "tx_hex" : "0100000001a246b6c5e5df9e98d0ce9dfc52b8cf0df9a6b1567790096411a5209e19e1e59000000000b6473044022055909a48a42fc424fb2b3ba590a1a798ca8e709047222d6c30c73b2f26fde73b02205bf63cf73483ddc191187135214d4a17008fbc4e0fc3961b200bc25b3cd9814d01514c6b6304f97e2968b175210315d9c51c657ab1be4ae9d3ab6e76a619d3bccfe830d5363fa168424c0d044732ac6782012088a914cd9e7193fbf9b6d75ae7b3026f394f55dd6f4b6c882103a27ba503d8d87636f289090f536d6c724ae76b16861e4241e07131fc52ab4025ac68feffffff01b04f0000000000001976a9141462c3dd3f936d595c9af55978003b27c250441f88ac8f669168"
   }
}
```
on chain: https://chainz.cryptoid.info/pivx/tx.dws?2d69c282ed1f59ade4f0d4f26ef2cdfa67aef29ad728835fddcf1424c27cd316.htm
fees were higher than output, but ok, nice

mariocynicys commented on mm2src/mm2_bitcoin/chain/src/block_header.rs:
```diff
@@ -230,16 +239,16 @@ impl Deserializable for BlockHeader {
             || reader.coin_variant().is_kmd_assetchain();
-        let hash_final_sapling_root = if is_zcash { Some(reader.read()?) } else { None };
+        let mut hash_final_sapling_root = if is_zcash_style { Some(reader.read()?) } else { None };
         let time = reader.read()?;
-        let bits = if is_zcash {
+        let bits = if is_zcash_style {
             BlockHeaderBits::U32(reader.read()?)
```
so is pivx a zcash_style or not?
- if it is, i see we are reading the hash_final_sapling_root before reading the nonce! (L244 & L251)
- if it is not, then we don't read the `solution` for it (L266)

shamardy replied:
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2572#issuecomment-3152977822

shamardy replied:
Also in the code there is this comment
```
// A PIVX header is a standard header with a `hash_final_sapling_root` added after the nonce.
```

mariocynicys replied:
okkie, pivx has no solution field.

mariocynicys reviewed (approved):
LGTM
